### PR TITLE
Nonlinear Hybrid

### PR DESCRIPTION
**Pull request by varunagrawal**
Add Nonlinear Hybrid Factor Graph machinery.

I moved all the tests for switching systems from our previous scheme to our current scheme and all of them pass!

### CODE REVIEW

File: gtsam/hybrid/HybridGaussianFactor.h [hunk 1]
```diff
@@ -37,6 +37,12 @@ class GTSAM_EXPORT HybridGaussianFactor : public HybridFactor {
   using This = HybridGaussianFactor;
   using shared_ptr = boost::shared_ptr<This>;
 
+  HybridGaussianFactor() = default;
+
+  HybridGaussianFactor(const KeyVector &continuousKeys,
+                       const DiscreteKeys &discreteKeys)
+      : Base(continuousKeys, discreteKeys) {}
```
Comment (xsj01) [inline gtsam/hybrid/HybridGaussianFactor.h:44]: I'm a bit confused why we could construct a Gaussian factor here without any parameters eg. A, b?
Also, it seems inner_ is empty by using this constructor.

File: gtsam/hybrid/HybridNonlinearFactorGraph.cpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      factors_[i]->print(ss.str(), keyFormatter);
+      std::cout << std::endl;
+    }
+  }
+}
+
+/* ************************************************************************* */
+bool HybridNonlinearFactorGraph::equals(const HybridNonlinearFactorGraph& other,
+                                        double tol) const {
+  return false;
```
Comment (xsj01) [inline gtsam/hybrid/HybridNonlinearFactorGraph.cpp:48]: Should it be return Base::equals(other, tol)?

**Review by xsj01 — APPROVED** [review]
LGTM with 2 minor comments

File: gtsam/hybrid/HybridGaussianFactorGraph.cpp [hunk 3]
```diff
@@ -98,6 +98,12 @@ GaussianMixtureFactor::Sum sumFrontals(
     } else if (f->isContinuous()) {
       deferredFactors.push_back(
           boost::dynamic_pointer_cast<HybridGaussianFactor>(f)->inner());
+
+    } else if (f->isDiscrete()) {
```
[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: Yeah there was one test that was failing without this. You can probably comment it out and find out which one.

**Comment by varunagrawal** [comment]
Folks, there are PRs before this one that need to be reviewed and merged first to make your lives easier. :)

File: gtsam/hybrid/HybridGaussianFactorGraph.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: I remember now. If we have discrete only factors it goes to the next condition and errors out since it is not an orphan. That shouldn't happen right? If we only have discrete factors left, we should just be returning a discrete decision tree and not erroring out.

@ProfFan can you please explain what was your rationale here?

File: gtsam/hybrid/HybridGaussianFactor.h (hunk 1 above)
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactor.h:44]: Ah good catch. Let me verify this really quick.

File: gtsam/hybrid/HybridNonlinearFactorGraph.cpp (hunk 2 above)
Comment (varunagrawal) [reply gtsam/hybrid/HybridNonlinearFactorGraph.cpp:48]: We can just get rid of this. :)

File: gtsam/hybrid/HybridGaussianFactorGraph.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ProfFan) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: I think because this is a GaussianMixtureFactor, there should not be any discrete factor added to it

[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: This is a Hybrid Factor Graph method so a discrete factor can be present.

[the remark this answers is not in the document]
Comment (ProfFan) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: I mean the return type is `GaussianMixtureFactor::Sum` (`DT<GFG>`), so this function should not process any discrete info?

[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: We're iterating over the factors in the graph so it's possible to have discrete factors. In that case we're not processing them, we just move on.

[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: @ProfFan can you give me the reason why discrete factors shouldn't be processed here? Without this update, we can't run elimination even on simple factor graphs as shown in the tests.

[the remark this answers is not in the document]
Comment (ProfFan) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: Because we eliminate discrete variables last, if any discrete variable appear here something is wrong

[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: If we see a discrete factor, it is just being skipped and not being considered in elimination. So there is no discrete factor in the GaussianMixture.

If we don't have this condition it goes to the next condition which throws an exception which is not correct.

Please see the whole function or better yet, try it out locally to get a better understanding.

[the remark this answers is not in the document]
Comment (ProfFan) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: It says `sumFrontals`, and when we do hybrid the frontals can **only** be continuous variables, so there should not be any discrete factors on it.

[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: Why not?

[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: And in that case what happens when the factor graph has for example a prior factor on a discrete variable?

[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: Fan, I don't think you're completely understanding this change and you don't seem to be reading my responses.
Moreover this is as much your project as mine so rather than provide cursory comments, please a look at this in detail and providing a resolution to the problem.

I have mentioned before that this change is necessary since a hybrid factor graph can have discrete only factors, and this check only serves to skip over discrete factors so we don't error out. Only hybrid or continuous factors are actually processed in this function.

[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: @ProfFan it's your original code that is looping over all the factors. Are you saying that when we loop over all the factors, we can't have discrete factors? That doesn't seem right to me.

[the remark this answers is not in the document]
Comment (ProfFan) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: I apologize if my comments look cursory to you. However I think because in the `HybridJunctionTree` special treatment has already been done to make sure that the arguments to `EliminateHybrid` will never have discrete factors. So if you use `eliminatePartialSequential` etc etc there is no problem.

[the remark this answers is not in the document]
Comment (ProfFan) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: The error only happens when you use `EliminateHybrid` directly. When I wrote this I my mental model is that `EliminateHybrid` is like an internal function and will not be called externally, so I did not consider this case in your unit test.

[the remark this answers is not in the document]
Comment (ProfFan) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: I am not opposing to adding this branching. I am a bit worried that people could start to use this function without thinking about the preconditions when this operation is valid. So I prefer to keep this function as-is, or add some comments in the function Doxygen as a warning.

[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: Okay now this makes sense. In that case, you could have mentioned this before rather than simply saying this is wrong. I am not aware of your mental model and there aren't any comments about the same.

I believe we should still test `EliminateHybrid` individually. Letting users know what the assumptions are is the right way here, but we should also ensure we handle these kind of edge cases so it makes using the API easier. Can you please comment what would you like me to add as a note in the docstring and I'll add that to this PR?

[the remark this answers is not in the document]
Comment (ProfFan) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: We will need to talk about the actual situation that will happen when this is called on an arbitrary factor graph. What I can think of is
```doxygen
Function to eliminate variables **under the following assumptions**:
1. When the ordering is fully continuous, and the graph only contains continuous and hybrid factors
2. When the ordering is fully discrete, and the graph only contains discrete factors
Any usage outside of this is incorrect.

\warning This function is not meant to be used with arbitrary hybrid factor graphs. For example, if there exists continuous parents, and one tries to eliminate a discrete variable (as specified in the ordering), the result will be INCORRECT and there will be NO error raised.
```

[the remark this answers is not in the document]
Comment (ProfFan) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: How does that sound to you?

[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp:102]: Sounds good to me.

File: gtsam/hybrid/HybridNonlinearFactorGraph.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        // In this case, we check if factor->inner() is nonlinear since
+        // HybridFactors wrap over continuous factors.
+        auto nlhf = boost::dynamic_pointer_cast<HybridNonlinearFactor>(factor);
+        if (auto nlf =
+                boost::dynamic_pointer_cast<NonlinearFactor>(nlhf->inner())) {
+          auto hgf = boost::make_shared<HybridGaussianFactor>(
+              nlf->linearize(continuousValues));
+          linearFG.push_back(hgf);
+        } else {
+          linearFG.push_back(factor);
```
Comment (ProfFan) [inline gtsam/hybrid/HybridNonlinearFactorGraph.cpp:79]: (ignore if not right) So we do allow pushing linear factors to the Hybrid NLFG? Because I remember that there is a `LinearContainerFactor`, not sure if we want to do the same here.

File: gtsam/hybrid/tests/testHybridIncremental.cpp [hunk 5]
```diff
@@ -0,0 +1,649 @@
...
+}
+
+/* ****************************************************************************/
+// // Test if we can approximately do the inference
+// TEST(DCGaussianElimination, Approx_inference) {
+//   Switching switching(4);
+//   IncrementalHybrid incrementalHybrid;
+//   HybridGaussianFactorGraph graph1;
+
+//   // Add the 3 DC factors, x1-x2, x2-x3, x3-x4
```
Comment (ProfFan) [inline gtsam/hybrid/tests/testHybridIncremental.cpp:234]: These are uncommented in the next PR, right?

File: gtsam/hybrid/tests/testHybridNonlinearFactorGraph.cpp [hunk 6]
```diff
@@ -0,0 +1,761 @@
...
+ */
+TEST(HybridsGaussianElimination, Eliminate_x2) {
+  Switching self(3);
+
+  // Gather factors on x2, will be two mixture factors (with x1 and x3, resp.).
+  HybridGaussianFactorGraph factors;
+  factors.push_back(self.linearizedFactorGraph[1]);  // involves m1
+  factors.push_back(self.linearizedFactorGraph[2]);  // involves m2
+
+  // TODO(Varun) remove this block since sum is no longer exposed.
```
Comment (ProfFan) [inline gtsam/hybrid/tests/testHybridNonlinearFactorGraph.cpp:277]: remove

**Review by ProfFan — APPROVED** [review]
Thanks Varun, looks great sans two minor comments!

File: gtsam/hybrid/tests/testHybridNonlinearFactorGraph.cpp (hunk 6 above)
Comment (varunagrawal) [reply gtsam/hybrid/tests/testHybridNonlinearFactorGraph.cpp:277]: Yeah I kept this since I wanted to highlight this change in the PR

File: gtsam/hybrid/HybridNonlinearFactorGraph.cpp (hunk 4 above)
Comment (varunagrawal) [reply gtsam/hybrid/HybridNonlinearFactorGraph.cpp:79]: Yes. This is something we can look at again once we update the type hierarchy for hybrid.

File: gtsam/hybrid/tests/testHybridIncremental.cpp (hunk 5 above)
Comment (varunagrawal) [reply gtsam/hybrid/tests/testHybridIncremental.cpp:234]: Yup
